### ISSUE TITLE
New unit test organization and build strategy

### ISSUE THREAD

**Issue by michaeltryby** [OpenWaterAnalytics/EPANET#427]
We have outgrown our current unit test organization and build strategy. We currently have eleven tests that get compiled into eleven separate test executables. Ten of these are for testing the epanet2 library and one is for testing the epanet-output library. 

The build is taking a long time because the Boost unit_test header file is very large and contains a lot of templates. We were compiling it 11 times. I have made some progress reorganizing things. 

I propose the following. We have one test module for the Toolkit. Within the test module we will have separate test suites for each of the functional categories found in the Toolkit API. The test suites will be in separate files and will be compiled together into a single test executable at build time. We will link with the boost testing library to eliminate some of the time spent compiling the test headers. 

These changes should help make things more manageable from both an organizational and build perspective.

**Comment by LRossman** [comment]
I have three questions:

1. Will we be able to run the individual test files in stand-alone fashion on our local machine, suitably modified to run without Boost (as I have currently been doing), to allow us to debug CI failures when we submit PRs?
2. Has the successful build status of the dev branch been restored or will we have to wait until the new unit test strategy is implemented before we can submit any new PRs?
3. Are there alternatives to using the Boost unit test library that may be worth exploring or are we too far along in the game for that?

**Comment by eladsal** [comment]
Good idea. If the combined test fails, will we know which part of it failed?

**Comment by michaeltryby** [comment]
@LRossman 
1) You are welcome to use your ingenuity to figure that out.
2) This is something I hope to check in later today or tomorrow.  
3) I suppose it's never to late, provided you are willing to rewrite all the tests :smile:

@eladsal Yes boost test tells you where the failures are.

**Comment by LRossman** [comment]
@michaeltryby you didn't really answer my first question. Let me rephrase it. Will the unit tests still exist as separate files in the `\tests` folder as they do now, which allows easy access to individual tests (and makes it easy to add new ones), or will they all be merged into a single file? I hope it's not the latter.
